pkg/apis/events/v1alpha1/channel_lifecycle.go

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2019 The Knative Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package v1alpha1
+
+import (
+	"knative.dev/pkg/apis"
+	"knative.dev/pkg/apis/duck/v1beta1"
+)
+
+// GetCondition returns the condition currently associated with the given type,
+// or nil.
+func (cs *ChannelStatus) GetCondition(t apis.ConditionType) *apis.Condition {
+	return channelCondSet.Manage(cs).GetCondition(t)
+}
+
+// IsReady returns true if the resource is ready overall.
+func (cs *ChannelStatus) IsReady() bool {
+	return channelCondSet.Manage(cs).IsHappy()
+}
+
+// InitializeConditions sets relevant unset conditions to Unknown state.
+func (cs *ChannelStatus) InitializeConditions() {
+	channelCondSet.Manage(cs).InitializeConditions()
+}
+
+// SetAddress updates the Addressable status of the channel and propagates a
+// url status to the Addressable status condition based on url.
+func (cs *ChannelStatus) SetAddress(url *apis.URL) {
+	if cs.Address == nil {
+		cs.Address = &v1beta1.Addressable{}
+	}
+	if url != nil {
+		cs.Address.URL = url
+		channelCondSet.Manage(cs).MarkTrue(ChannelConditionAddressable)
+	} else {
+		cs.Address.URL = nil
+		channelCondSet.Manage(cs).MarkFalse(ChannelConditionAddressable, "emptyUrl", "url is empty")
```

Comment:
Should there be a check for `url != nil && url.String() == ""`?

Reply:
there is a fair amount of trust at this point of the app and if the controller wants to set a "" url, I guess that is its choice, I don't return an error here so there is no way to signal back up the call that this was odd.